tasks/migrate-removals.ts

```diff
@@ -0,0 +1,252 @@
+import { getRemoval } from '../utils/contracts';
+
+interface MigrateRemovalsTaskOptions {
+  file: string;
+  outputFile?: string;
+  dryRun?: boolean;
+}
+
+type ParsedMigrateRemovalsTaskOptions = RequiredKeys<
+  MigrateRemovalsTaskOptions,
+  'file'
+>;
+
+const asciiStringToHexString = (ascii: string): string => {
+  return `0x${[...Array.from({ length: ascii.length }).keys()]
+    .map((index) => ascii.charCodeAt(index).toString(16))
+    .join('')}`;
+};
+
+export const GET_MIGRATE_REMOVALS_TASK = () =>
+  ({
+    name: 'migrate-removals',
+    description: 'Utility to mint legacy removals',
+    run: async (
+      options: MigrateRemovalsTaskOptions,
+      _: CustomHardHatRuntimeEnvironment
+    ): Promise<void> => {
+      const {
+        file,
+        outputFile = 'migrated-removals.json',
+        dryRun,
+      } = options as ParsedMigrateRemovalsTaskOptions;
+      const jsonData = JSON.parse(readFileSync(file, 'utf8'));
+      // hre.log({ jsonData });
+
+      const [signer] = await hre.getSigners();
+      const signerAddress = await signer.getAddress();
+      const removalContract = await getRemoval({
+        hre,
+        signer,
+      });
+      hre.log(`Removal contract address: ${removalContract.address}`);
+      hre.log(`Signer address: ${signerAddress}`);
+      // const fireblocksSigner = removalContract.signer as FireblocksSigner;
+
+      if (!dryRun) {
+        hre.log(
+          chalk.bold.white(
+            `✨ Minting removals for ${jsonData.length} projects...`
+          )
+        );
+        hre.log(chalk.white(`🤞 Submitting transactions...`));
+      } else {
+        hre.log(
+          chalk.bold.white(
+            `DRY RUN 🌵 Minting removals for ${jsonData.length} projects...`
+          )
+        );
+      }
+      // submit all mintBatch transactions serially to avoid nonce collision!
+      const pendingTransactions = [];
+      let projectIndex = 1;
+      for (const project of jsonData) {
+        const amounts = project.amounts.map((amount: string) =>
+          ethers.utils
+            .parseUnits(divide(Number(amount), 1_000_000).toString())
+            .toString()
+        );
+
+        const removals = project.removals.map((removal: any, index) => {
+          const removalData = {
+            idVersion: Number(removal.idVersion),
+            // methodology: Number(removal.methodology), // TODO
+            // methodologyVersion: Number(removal.methodologyVersion), // TODO
+            methodology: 1,
+            methodologyVersion: 0,
+            vintage: Number(removal.vintage),
+            country: asciiStringToHexString(removal.country),
+            subdivision: asciiStringToHexString(
+              removal.subdivision.slice(0, 2)
+            ), // TODO we only want the state code here, original script was giving full substring
+            supplierAddress: '0x9A232b2f5FbBf857d153Af8B85c16CBDB4Ffb667', // TODO need real supplier address on the project
```

Comment:
fix before merging

Reply:
definitely will do - leaving this here just for now since otherwise we're going to have to toggle this back and forth for testing with data that doesn't have this value (none of our prod data actually has wallet addresses for suppliers... we should probably handle testing that by modifying the input file or something though)

Reply:
I think it's best to merge code as close to the intended state as possible, and anything that's convenience for the sake of future testing should just be commented out for the time being. So I'd suggest commenting this and the sub-identifier out, and replace them with `removal. supplierAddress` and `removal.subIdentifier`.
Then when testing locally we can uncomment those out and leave it like that locally.